Improve error messages for syntax errors

rhuanjl opened the pull request:
Picking up on discussion in #5178 and also issue #767

This PR:
1. Adds support for %s in syntax error messages (though only up to 2 uses of it per message)
1. Adds facility for getting a string representation of the token being parsed, and the previous token
1. Updates some commonly encountered error messages that currently only say "Syntax error"

I'd like to follow up and improve some other error messages but this seemed big enough for one PR.

I'm slightly nervous that I may have done something wrong in my implementation of %s support - it works in my tests but I don't 100% understand all the relevant logic.

fixes #5178

dilijev commented on lib/Parser/perrors.h:
```diff
@@ -109,3 +109,10 @@ LSC_ERROR_MSG(1093, ERRLabelBeforeClassDeclaration, "Labels not allowed before c
+LSC_ERROR_MSG(1201, ERRTokenAfter, "Unexpected token '%s' after '%s'")
+LSC_ERROR_MSG(1202, ERRIdentifierAfter, "Unexpected identifier '%s' after '%s'")
+LSC_ERROR_MSG(1203, ERRInvalidIdentifier, "Unexpected invalid identifier '%s' after '%s'")
+LSC_ERROR_MSG(1205, ERRValidIfFollowedBy, "%s is only valid if followed by %s")
```
Maybe quotes around these `%s` for consistency?

dilijev replied:
Nevermind, I see how this message is being used now. LGTM

dilijev commented on test/es6/HTMLComments.js:
```diff
@@ -22,7 +22,8 @@ WScript.Echo("Code before CRLF--> is reachable");
-WScript.Echo("Code before CR--> is reachable");
--> WScript.Echo("Code after CR--> is unreachable");
+WScript.Echo("Code before CR--> is reachable");
+--> WScript.Echo("Code after CR--> is unreachable");
```
These carriage return (CR, `\r`, or `0x0d`) characters were deliberately placed here without newline (`\n` or `0x0a`) and are part of the parser test. Can you revert these changes?

rhuanjl replied:
I don't recall editing that line possibly an oddity of my vscode settings when I made the change further down - will revert with a hex editor.

dilijev replied:
Yeah this is one of those things that editors will generally try to make sane. Getting the file into this weird line ending state is almost a matter of just editing the bytes.

dilijev commented on test/es6/HTMLComments.js:
```diff
@@ -39,10 +40,12 @@ WScript.Echo("Code before PS--> is reachable"); --> WScript.Echo("Code after P
-WScript.Echo("Code before CRPS--> is reachable");
 --> WScript.Echo("Code after CRPS--> is unreachable");
+WScript.Echo("Code before CRPS--> is reachable");
+ --> WScript.Echo("Code after CRPS--> is unreachable");
```
(same as above)

dilijev commented on test/es6/HTMLComments.js:
```diff
@@ -39,10 +40,12 @@ WScript.Echo("Code before PS--> is reachable"); --> WScript.Echo("Code after P
-WScript.Echo("Code before CRLS--> is reachable");
 --> WScript.Echo("Code after CRLS--> is unreachable");
+WScript.Echo("Code before CRLS--> is reachable");
+ --> WScript.Echo("Code after CRLS--> is unreachable");
```
(same as above)

dilijev commented on test/es6/HTMLComments.js:
```diff
@@ -71,4 +74,4 @@ assert.throws(function () { eval('/* */--> WScript.Echo("Code after /* */--> is
 assert.areEqual(0, a, "Post decrement executes");
 
-assert.throws(function () { eval('/* */ --->'); }, SyntaxError, "HTMLCloseComment causes syntax error with an extra -", "Syntax error");
+assert.throws(function () { eval('/* */ --->'); }, SyntaxError, "HTMLCloseComment causes syntax error with an extra -", "Unexpected token '>' after '-'");
```
These error messages are perhaps more descriptive of the parser state but probably just as confusing to the author. IMO this is fine though.

rhuanjl replied:
Yeah I didn't like that one much but wasn't sure what to do with it; the parser has -- i.e. predecrement operator then a - i.e. unary minus then is expecting a statement but hits > I could special case it but would be a bit messy and wasn't sure it was worth it. At least with this option the author knows that it's the "-" and ">" that there's a problem with rather than looking earlier/later in the line.

dilijev replied:
Probably not worth it. I think this is fine.

rhuanjl commented:
Two things I didn't highlight in my description above:
1. I intentionally left a gap in the error numbers in perrors.h before the ones I added with a thought that these more generic errors should stay at the bottom - hence leaving space for new specific ones before them. In hindsight not sure why I didn't put a comment in for this.
2. There are 4 calls to processError that I haven't updated to handle the parameters for %s they're in ScriptContext.cpp, BytecodeGenerator.cpp and SourceTextModuleRecord.cpp. I couldn't work out if parse errors could reach these points - if yes I need to update them or the error messages won't show correctly for those paths.

dilijev commented:
> 2\. if yes I need to update them or the error messages won't show correctly for those paths.

IMO We should assume this can happen (otherwise it's dead code, right?) and get those messages updated.

rhuanjl commented:
I have added the comment I mentioned above to perrors.h and updated to latest.

And I've re-examined all calls to ProcessException() and concluded that the others don't need updating to handle %s.
These are:
1. In SourceTextModuleRecord.cpp: Used specifically to handle OutOfMemory and StackOverFlow during parsing but not syntax errors which follow a different path.
2. In ByteCodeGenerator.cpp: Used for exceptions whilst generating byte code, not syntax errors.
3. 2 Calls in ScriptContext.cpp used in SerialiseScript and SerializeParserState both of these don't appear to be designed to handle syntax errors but rather out of memory and stackoverflow - though there's possibly a logical error here - when you hit a syntax error in SerializeParserState whilst testing with ch and -UseParserStateCache you get two error messages printed (one meaningful and one not) - either way I think this is a different issue.

Good to merge?

dilijev commented on lib/Parser/cmperr.cpp:
```diff
@@ -4,8 +4,16 @@
-ParseExceptionObject::ParseExceptionObject(HRESULT hr) : m_hr(hr)
+ParseExceptionObject::ParseExceptionObject(HRESULT hr, LPCWSTR stringOneIn, LPCWSTR stringTwoIn)
 {
+    m_hr = hr;
```
why move this assignment from the constructor-initialization syntax to the body of the constructor?

rhuanjl replied:
That wasn't a new change. There was no real reason for it though, I thought it looked neater next to the two string properties, want me to revert it?

dilijev replied:
I don't **think** it makes any difference. I guess we don't have a code guideline or real consistency around this point AFAICT. @Penguinwizzard do you know if this matters in terms of code gen?

You can leave it alone for now, so we can let the CI finish and merge this if it doesn't make a difference.

Penguinwizzard replied:
It's mostly stylistic, and there's no real enforced style in ChakraCore (yet, anyway). I'd be decently surprised if we get different codegen due to this in optimized builds.

dilijev commented:
I'm running internal verification tests and will merge when they complete.

dilijev commented:
Some internal test baselines affected by the change of error messages need updating. I'll take care of that and then merge the PR. Thanks @rhuanjl!
